### DISCUSSION TITLE
API to enumerate open binary views/auto-register notification listeners

### DISCUSSION THREAD

**Discussion by 404d** [Vector35/binaryninja-api#1577] [General] [answered]
Is there a good way for a plugin to enumerate/get notified about new binary views opened by the user?  I remember looking into this some time ago but wasn't able to find a non-hacky way that worked well.

I think it might be slightly more doable nowadays by for example abusing the UI API, but this is probably something there should be a proper API for.

---

I have mainly two usecases for this:
- **Cross-BV plugins**
  
  [`peutils`](https://github.com/404d/peutils) for example copies subroutine names and type definitions between PE BVs based on export/import relations.

- **Global `BinaryDataNotification` listeners**

  I have a private plugin that auto-guesses array length of some structs for a proprietary library, by listening to `data_var_added` and `data_var_updated`, looking for non-array declarations of these structs.
  
  Someone also posted the following example to the public Slack recently, which renames string data vars based on the string value. This plugin could be automated by listening for events in the same way as described above.
  https://binaryninja.slack.com/archives/C0CVALTLN/p1587205479137200
  
  Both examples are scenarios where it could be relevant for the plugin to auto-register for notifications for all views.

Currently, I've implemented both use cases by registering context menu actions that must be executed on each individual binary view.

**Comment by psifertex** [comment, marked as answer]
This is useful functionality that we agree would be helpful. Sorry for not replying yet, but here's the list of issues you can up-vote/subscribe to to track when the functionality is complete:
#1097 , specifically number 5. Additionally, a related feature that might also be useful is #505

**Comment by rssor** [comment]
And, specifically, we've talked a lot about adding an API specifically for a callback that runs when BinaryViews are created, but don't have an issue for it yet, so slightly different from the 1097/505 issues.
